kvbc/test/categorization/versioned_kv_category_unit_test.cpp

```diff
@@ -0,0 +1,625 @@
+  {
+    auto values = std::vector<std::optional<categorization::Value>>{};
+    cat.multiGet({"k"}, {1}, values);
+    const auto expected = std::vector<std::optional<categorization::Value>>{VersionedValue{{1, "v1"}}};
+    ASSERT_EQ(values, expected);
+  }
+
+  // Key "k" is no longer available for blocks >= 2 via get().
+  {
+    ASSERT_FALSE(cat.get("k", 2));
+    ASSERT_FALSE(cat.get("k", 3));
+  }
+
+  // Key "k" is no longer available for blocks >= 2 via multiGet().
+  {
+    auto values = std::vector<std::optional<categorization::Value>>{};
+    cat.multiGet({"k", "k"}, {2, 3}, values);
+    const auto expected = std::vector<std::optional<categorization::Value>>{std::nullopt, std::nullopt};
+    ASSERT_EQ(values, expected);
+  }
+
+  // Key "k" has no latest value via getLatest().
+  { ASSERT_FALSE(cat.getLatest("k")); }
+
+  {
+    // Key "k" has no latest value via multiGetLatest().
+    auto values = std::vector<std::optional<categorization::Value>>{};
+    cat.multiGetLatest({"k"}, values);
+    const auto expected = std::vector<std::optional<categorization::Value>>{std::nullopt};
+    ASSERT_EQ(values, expected);
+  }
+
+  // Key "k" has a latest version of 2 via getLatestVersion() - the version it was deleted at.
+  {
+    const auto latest_version = cat.getLatestVersion("k");
+    ASSERT_TRUE(latest_version);
+    ASSERT_EQ(*latest_version, 2);
+  }
+
+  // Key "k" has a latest version of 2 via multiGetLatestVersion() - the version it was deleted at.
+  {
+    auto latest_versions = std::vector<std::optional<BlockId>>{};
+    cat.multiGetLatestVersion({"k"}, latest_versions);
+    const auto expected = std::vector<std::optional<BlockId>>{2};
+    ASSERT_EQ(latest_versions, expected);
+  }
+}
+
+TEST_F(versioned_kv_category, propagate_stale_on_update) {
+  {
+    const auto stale_on_update = false;
+    auto in = VersionedInput{};
+    in.calculate_root_hash = true;
+    in.kv["k"] = ValueWithFlags{"v1", stale_on_update};
+    const auto out = add(1, std::move(in));
+
+    ASSERT_FALSE(out.keys.cbegin()->second.stale_on_update);
+  }
+
+  {
+    const auto stale_on_update = true;
+    auto in = VersionedInput{};
+    in.calculate_root_hash = true;
+    in.kv["k"] = ValueWithFlags{"v2", stale_on_update};
+    const auto out = add(2, std::move(in));
+
+    ASSERT_TRUE(out.keys.cbegin()->second.stale_on_update);
+  }
+}
+
+TEST_F(versioned_kv_category, delete_genesis) {
+  const auto mark_stale_on_update = true;
+  const auto non_stale_on_update = false;
+
+  // Add keys "ka", "kb", "kc" and "kd" in block 1. Mark "kc" as stale-on-update.
+  auto out1 = VersionedOutput{};
+  {
+    auto in = VersionedInput{};
+    in.calculate_root_hash = true;
+    in.kv["ka"] = ValueWithFlags{"va1", non_stale_on_update};
+    in.kv["kb"] = ValueWithFlags{"vb1", non_stale_on_update};
+    in.kv["kc"] = ValueWithFlags{"vc1", mark_stale_on_update};
+    in.kv["kd"] = ValueWithFlags{"vd1", mark_stale_on_update};
```

Comment:
Not sure if this was intended to be stale_on_update based on the comment.

Reply:
I will change the comment to include "kd" too, thanks!